File: ipcrypt.py
#!/usr/bin/env python

import csv
import struct
import sys

# copy your key here
KEY = 'some 16-byte key'
# CSV delimiter character
DELIMITER = ','  


def rotl(b, r):
    return ((b << r) & 0xff) | (b >> (8 - r))


def permute_fwd(state):
    (b0, b1, b2, b3) = state
    b0 += b1
    b2 += b3
    b0 &= 0xff
    b2 &= 0xff
    b1 = rotl(b1, 2)
    b3 = rotl(b3, 5)
    b1 ^= b0
    b3 ^= b2
    b0 = rotl(b0, 4)
    b0 += b3
    b2 += b1
    b0 &= 0xff
    b2 &= 0xff
    b1 = rotl(b1, 3)
    b3 = rotl(b3, 7)
    b1 ^= b2
    b3 ^= b0
    b2 = rotl(b2, 4)
    return (b0, b1, b2, b3)


def permute_bwd(state):
    (b0, b1, b2, b3) = state
    b2 = rotl(b2, 4)
    b1 ^= b2
    b3 ^= b0
    b1 = rotl(b1, 5)
    b3 = rotl(b3, 1)
    b0 -= b3
    b2 -= b1
    b0 &= 0xff
    b2 &= 0xff
    b0 = rotl(b0, 4)
    b1 ^= b0
    b3 ^= b2
    b1 = rotl(b1, 6)
    b3 = rotl(b3, 3)
    b0 -= b1
    b2 -= b3
    b0 &= 0xff
    b2 &= 0xff
    return (b0, b1, b2, b3)


def xor4(x, y):
    return [(x[i] ^ y[i]) & 0xff for i in (0, 1, 2, 3)]


def encrypt(key, ip):
    """16-byte key, ip string like '192.168.1.2'"""
    k = [struct.unpack('<B', x)[0] for x in key]
    try:
        state = [int(x) for x in ip.split('.')]
    except ValueError:
        raise
    try:
        state = xor4(state, k[:4])
        state = permute_fwd(state)
        state = xor4(state, k[4:8])
        state = permute_fwd(state)
        state = xor4(state, k[8:12])
        state = permute_fwd(state)
        state = xor4(state, k[12:16])
    except IndexError:
        raise
    return '.'.join(str(x) for x in state)


def decrypt(key, ip):
    """16-byte key, encrypted ip string like '215.51.199.127'"""
    k = [struct.unpack('<B', x)[0] for x in key]
    try:
        state = [int(x) for x in ip.split('.')]
    except ValueError:
        raise
    try:
        state = xor4(state, k[12:16])
        state = permute_bwd(state)
        state = xor4(state, k[8:12])
        state = permute_bwd(state)
        state = xor4(state, k[4:8])
        state = permute_bwd(state)
        state = xor4(state, k[:4])
    except IndexError:
        raise
    return '.'.join(str(x) for x in state)


def usage():
    print 'usage:  %s csvfile index e|d'  % sys.argv[0]
    print '\tindex = csv index in 0, 1, ...'
    print '\te = encrypt, d = decrypt'
    sys.exit(0)


def test():
    """basic encryption sanity check"""
    ip = init = '1.2.3.4'
    key = '\xff'*16 
    iterations = 10
    for i in xrange(iterations):
        ip = encrypt(key, ip)
    if ip != '191.207.11.210':
        raise ValueError
    for i in xrange(iterations):
        ip = decrypt(key, ip)
    if ip != init:
        raise ValueError


def main():
    assert len(KEY) == 16
    try:
        test()
    except ValueError:
        print 'test failed'
        sys.exit(0)

    try:
        filein = sys.argv[1]
        index = int(sys.argv[2])
        mode = sys.argv[3]
    except:
        usage()

    if mode == 'e':
        process = encrypt
    elif mode == 'd':
        process = decrypt
    else:
        usage()

    with open(filein, 'rb') as csvfile:
        reader = csv.reader(csvfile, delimiter=DELIMITER)
        writer = csv.writer(sys.stdout, delimiter=DELIMITER)

        for row in reader:
            ip = row[index].strip()
            newrow = row
            try:
                newrow[index] = process(KEY, ip)
            except:
                continue
            writer.writerow(newrow)


if __name__ == '__main__':
    sys.exit(main())

File: ipcrypt.go
package main

// see also https://github.com/dgryski/go-ipcrypt/

import (
	"encoding/csv"
	"errors"
	"fmt"
	"io"
	"net"
	"os"
	"strconv"
	"strings"
)

// copy your key here
const (
	KEY = "some 16-byte key"
)

func rotl(x, c byte) byte {
	return (x << c) | (x >> (8 - c))
}

func permute_fwd(state [4]byte) [4]byte {
	a := state[0]
	b := state[1]
	c := state[2]
	d := state[3]
	a += b
	c += d
	a &= 0xff
	c &= 0xff
	b = rotl(b, 2)
	d = rotl(d, 5)
	b ^= a
	d ^= c
	a = rotl(a, 4)
	a += d
	c += b
	a &= 0xff
	c &= 0xff
	b = rotl(b, 3)
	d = rotl(d, 7)
	b ^= c
	d ^= a
	c = rotl(c, 4)
	return [4]byte{a, b, c, d}
}

func permute_bwd(state [4]byte) [4]byte {
	a := state[0]
	b := state[1]
	c := state[2]
	d := state[3]
	c = rotl(c, 4)
	b ^= c
	d ^= a
	b = rotl(b, 5)
	d = rotl(d, 1)
	a -= d
	c -= b
	a &= 0xff
	c &= 0xff
	a = rotl(a, 4)
	b ^= a
	d ^= c
	b = rotl(b, 6)
	d = rotl(d, 3)
	a -= b
	c -= d
	a &= 0xff
	c &= 0xff
	return [4]byte{a, b, c, d}
}

func xor4(x [4]byte, y []byte) [4]byte {
	return [4]byte{x[0] ^ y[0], x[1] ^ y[1], x[2] ^ y[2], x[3] ^ y[3]}
}

func bytes2ip(bytes [4]byte) string {
	ipaddr := []string{"", "", "", ""}
	ipaddr[0] = strconv.Itoa(int(bytes[0]))
	ipaddr[1] = strconv.Itoa(int(bytes[1]))
	ipaddr[2] = strconv.Itoa(int(bytes[2]))
	ipaddr[3] = strconv.Itoa(int(bytes[3]))
	return strings.Join(ipaddr, ".")
}

func Encrypt(k [16]byte, ip string) (string, error) {
	p := net.ParseIP(ip)
	if p == nil {
		return "", errors.New("encrypt: invalid IP")
	}
	state := [4]byte{p[12], p[13], p[14], p[15]}

	state = xor4(state, k[:4])
	state = permute_fwd(state)
	state = xor4(state, k[4:8])
	state = permute_fwd(state)
	state = xor4(state, k[8:12])
	state = permute_fwd(state)
	state = xor4(state, k[12:16])

	return bytes2ip(state), nil
}

func Decrypt(k [16]byte, ip string) (string, error) {
	p := net.ParseIP(ip)
	if p == nil {
		return "", errors.New("encrypt: invalid IP")
	}
	state := [4]byte{p[12], p[13], p[14], p[15]}

	state = xor4(state, k[12:16])
	state = permute_bwd(state)
	state = xor4(state, k[8:12])
	state = permute_bwd(state)
	state = xor4(state, k[4:8])
	state = permute_bwd(state)
	state = xor4(state, k[:4])

	return bytes2ip(state), nil
}

func test() error {
	ip := "1.2.3.4"
	init := ip
	var err error
	var key [16]byte
	for i := 0; i < 16; i++ {
		key[i] = 0xff
	}
	for i := 0; i < 10; i++ {
		ip, err = Encrypt(key, ip)
		if err != nil {
			return err
		}
	}
	if ip != "191.207.11.210" {
		return errors.New("test failed: wrong intermediate value")
	}
	for i := 0; i < 10; i++ {
		ip, err = Decrypt(key, ip)
		if err != nil {
			return err
		}
	}
	if init != ip {
		return errors.New("test failed: decrypted values doesn't match")
	}
	return nil
}

func main() {
	err := test()
	if err != nil {
		fmt.Println(err)
		return
	}
	if len(os.Args) < 4 {
		fmt.Println("not enough arguments")
		return
	}
	filein := os.Args[1]
	index, err := strconv.Atoi(os.Args[2])
	if err != nil {
		fmt.Println(err)
		return
	}
	mode := os.Args[3]

	var process func([16]byte, string) (string, error)
	if mode == "e" {
		process = Encrypt
	} else if mode == "d" {
		process = Decrypt
	} else {
		fmt.Println("error: wrong mode")
		return
	}

	file, err := os.Open(filein)
	if err != nil {
		fmt.Println(err)
		return
	}

	defer file.Close()

	reader := csv.NewReader(file)
	reader.Comma = ','
	writer := csv.NewWriter(os.Stdout)
	writer.Comma = ','

	var key [16]byte
	for i := 0; i < 16; i++ {
		key[i] = byte(KEY[i])
	}

	for {
		line, err := reader.Read()
		if err == io.EOF {
			break
		} else if err != nil {
			fmt.Println(err)
			return
		}
		newline := line
		newline[index], err = process(key, line[index])
		if err != nil {
			fmt.Println(err)
			return
		}
		if err == nil {
			writer.Write(newline)
		}
	}

	writer.Flush()
}
